storage: enable switching env through encryption flags

mberhault opened the pull request:
Part of encryption work. See [Encryption RFC](https://github.com/cockroachdb/cockroach/pull/19785)

Add --enterprise-encryption flag and match it to stores.
Enabling encryption forces the use of the "Switching Env"
store version.

If encryption is not requested, we do not bump the version. This is
to allow downgrades to older binaries that do not support the switching
env. This will be changed after "a few" releases.

Commit:
storage: add support for encryption flags.

encryption-at-rest work.

Add --enterprise-encryption flag and match it to stores.
Enabling encryption forces the use of the "Switching Env"
store version.

If encryption is not requested, we do not bump the version. This is
to allow downgrades to older binaries that do not support the switching
env. This will be changed after "a few" releases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20225)
<!-- Reviewable:end -->

mberhault commented:
Pulled in some commits from other WIP (and closed others). See commit descriptions.

mberhault commented:
I'd appreciate a look, I need this functionality to keep going on the libroach side.

tschottdorf commented:
Looking, sorry for the delay!

mberhault commented on pkg/base/store_spec.go:
```diff
@@ -40,17 +40,37 @@ import (
+	// EncryptionSpec is non-nil after MatchStoreAndEncryptionSpecs if there is an encryption
+	// spec with matching path.
+	//	EncryptionSpec *StoreEncryptionSpec
+	ExtraFields map[string]string
```
btw: this clearly needs to go, it's not like we can easily pass a string->string map to cgo, and even if we could, that's a bit fragile.
I'm still trying to figure out if I can have some blind `CCLOptions` of some type, but it's a bit tricky.

tschottdorf commented:
Looks good so far! The meat of my comments are predictably in the area of ccl-oss glueing. I'll hold off on looking at the `ExtraFields` until prompted otherwise.

The C++ looked mostly trivial so far, but if there's any subtlety in then, I have missed it. I'm also not an experienced C++ programmer (or one at all).

---

Reviewed 9 of 9 files at r1, 7 of 7 files at r2, 13 of 13 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/db.cc, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-8yvExVJ2GAyq7TEjE:-L-8yvExVJ2GAyq7TEjF:bpuv4tc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f01fd8006ecacd715799e94f174d3010cb0943f9/c-deps/libroach/db.cc#L101)):*
> ```cpp
>   // Construct a new DBImpl from the specified DB and Env. Both the DB
>   // and Env will be deleted when the DBImpl is deleted. It is ok to
>   // pass NULL for the Env.
> ```

Comment needs an update now that there are two envs. What is the role of each?

---

*[c-deps/libroach/env_switching.cc, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-8zdsu8Kw3w3JXibIQ:-L-8zdsu8Kw3w3JXibIR:b-ro5pi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f01fd8006ecacd715799e94f174d3010cb0943f9/c-deps/libroach/env_switching.cc#L22)):*
> ```cpp
> //                 |        |
> //                 |   encrypted_env (EnvWrapper)
> //                 |        |
> ```

```
\        /
  \    /
    \/
base_env
   (Env)
```

and a similar cone at the top. I had to scratch my head for a couple dozen seconds.

---

*[c-deps/libroach/env_switching.cc, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-8zWxV-xQkh6BD1ljI:-L-8zWxV-xQkh6BD1ljJ:biul1lu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f01fd8006ecacd715799e94f174d3010cb0943f9/c-deps/libroach/env_switching.cc#L30)):*
> ```cpp
> };
> 
> rocksdb::Env* NewSwitchingEnv(rocksdb::Env* base_env) {
> ```

Do we ever call this without an env? If not, probably safer to disallow it.

---

*[pkg/base/store_spec.go, line 69 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-910L3fwREQs8Tf1KC:-L-910L3fwREQs8Tf1KD:bc1g06c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/base/store_spec.go#L69)):*
> ```Go
> 	InMemory    bool
> 	Attributes  roachpb.Attributes
> 	// EncryptionSpec is non-nil after MatchStoreAndEncryptionSpecs if there is an encryption
> ```

What's happening here?

---

*[pkg/base/store_spec.go, line 170 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-91APOh7DHOUGeut9E:-L-91APOh7DHOUGeut9F:b1wj8p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/base/store_spec.go#L170)):*
> ```Go
> 		}
> 
> 		switch field {
> ```

```go
const path = "path"
```

and then use that twice below.

---

*[pkg/base/store_spec.go, line 69 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-94TNURzRtP3h4i_i0:-L-94TNVTjldz2nFdH0Q:b74x5tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L69)):*
> ```Go
> 	InMemory        bool
> 	Attributes      roachpb.Attributes
> 	UseSwitchingEnv bool
> ```

Needs explanation.

---

*[pkg/base/store_spec.go, line 73 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-7IR0AC778-MlT7fQq-r4-73:-L-913pJqPDh4tul1Ps8:b2bybz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L73)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

btw: this clearly needs to go, it's not like we can easily pass a string->string map to cgo, and even if we could, that's a bit fragile.
I'm still trying to figure out if I can have some blind `CCLOptions` of some type, but it's a bit tricky.
</blockquote></details>

Ack, I'll just ignore this part for now.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 39 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-91UmvKHphnYcZ5QVj:-L-91Umwa-63gtw4AB_A:bqx8yvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/baseccl/encryption_spec.go#L39)):*
> ```Go
> func (es StoreEncryptionSpec) String() string {
> 	// All fields are set.
> 	return fmt.Sprintf("path=%s,key=%s,old-key=%s,rotation-period=%s",
> ```

Why not https://golang.org/pkg/net/url/#URL and avoid a hand-rolled format (simply strip the `<Scheme>://`)? That avoids lots of subtleties with the characters allowed in paths and looks familiar to everyone.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 129 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-927iRr5S9iJMagc3i:-L-927iRr5S9iJMagc3j:b-az27gw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/baseccl/encryption_spec.go#L129)):*
> ```Go
> 	var buffer bytes.Buffer
> 	for _, ss := range encl.Specs {
> 		fmt.Fprintf(&buffer, "--enterprise-encryption=%s ", ss)
> ```

I'm confused why this returns the hardcoded flag name.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 161 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-94eZb74edHDAdP_P2:-L-94eZb74edHDAdP_P3:b-4kt0jo) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/ccl/baseccl/encryption_spec.go#L161)):*
> ```Go
> // MatchStoreAndEncryptionSpecs iterates through the StoreEncryptionSpecList and looks
> // for matching paths in the StoreSpecList.
> // Any unmatched StoreEncryptionSpec causes an error.
> ```

The comment doesn't mention that we're going to manipulate the store spec list. Should this be called `MergeEncryptionSpecsIntoStoreSpecs`?

---

*[pkg/ccl/cliccl/start.go, line 40 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-932nl6bmtBybka5kV:-L-932nl6bmtBybka5kW:b-lqesd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/cliccl/start.go#L40)):*
> ```Go
> 	// This relies upon init-order between packages.
> 	cli.StartCmd.Flags().VarP(&storeEncryptionSpecs, "enterprise-encryption", "", encryptionFlagDesc)
> 	cli.ExtraStartPreRunHook = matchStoreEncryptionSpecs
> ```

Just do

```go
wrapped := cli.StartCmd.PersistentPreRunE
cli.StartCmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
  if err := wrapped(cmd, args); err != nil {
    return err
  }
  return matchStoreEncryptionSpecs(cmd, args)
}
```

and there is no need for `ExtraStartPreRunHook`.

To make this independent of init order, you can make both setters to `PersistentPreRunE` use this pattern.

---

*[pkg/ccl/cliccl/start.go, line 46 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-949-V1gGLQQkdpW7A:-L-949-V1gGLQQkdpW7B:bre2pku) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/cliccl/start.go#L46)):*
> ```Go
> // parsed stores.
> func matchStoreEncryptionSpecs() error {
> 	return baseccl.MatchStoreAndEncryptionSpecs(cli.ServerCfg.Stores, storeEncryptionSpecs)
> ```

Instead of exposing all of `ServerCfg`, just expose a getter into `serverCfg.Stores()` from `cli` that is clearly marked as something you shouldn't use in other contexts.

---

*[pkg/cli/context.go, line 61 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-93y5SVGzHL6PA4d_Y:-L-93y5SVGzHL6PA4d_Z:b-g6pb1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/cli/context.go#L61)):*
> ```Go
> }
> 
> // ServerCfg is the config for the `start` command.
> ```

Not exclusively, I believe.

---

*[pkg/cli/start.go, line 654 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-94KT7ksT2cYdIYuzs:-L-94KT7ksT2cYdIYuzt:bq8vm2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/cli/start.go#L654)):*
> ```Go
> 				if spec.ExtraFields != nil {
> 					for k, v := range spec.ExtraFields {
> 						fmt.Fprintf(tw, "  %s:\t%s\n", k, v)
> ```

Ignoring for now.

---

*[pkg/storage/engine/enginepb/registry.proto, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-90Laewu-DQp83U2iR:-L-90Laewu-DQp83U2iS:baewvbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fa0a4df33dd20e17eade406d3f2582970147095/pkg/storage/engine/enginepb/registry.proto#L40)):*
> ```protobuf
> 
> message FileEntry {
>   string Filename = 1;
> ```

Add a comment specifying if this is a relative or absolute path. I suppose it's relative (to what?) to allow moving of mount points.
There is also the concern of moving a HDD from a Linux to a Windows machine. Will things break because the path separator changes? It might be better to store a repeated string with individual path components.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-:-L-95B3_jJYB_79GqmK9:b-3y90x)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Thanks for the review. I'll loop back around when I've sorted out how to properly pass CCL options through rocksdb, most likely not the `ExtraFields`.

The only trick in the C++ code is who owns what for deletion. The envs aren't actually used by `DBImpl`, they're only held as scoped pointers to delete when `DBImpl` is deleted. This will change around quite a bit in future encryption 
work.

And if it makes you feel better: I don't know any experienced C++ programmers, just gray-haired ones.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/db.cc, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-8yvExVJ2GAyq7TEjE:-L-9GylR82CMscsjEmnN:bpjtqdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f01fd8006ecacd715799e94f174d3010cb0943f9/c-deps/libroach/db.cc#L101)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Comment needs an update now that there are two envs. What is the role of each?
</blockquote></details>

Done. I still need to figure out exactly who will own what. Right now, they're only passed to the `DBImpl` to be deleted properly. They're only used by the parent `DBEngine` through the options.

---

*[c-deps/libroach/env_switching.cc, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-8zdsu8Kw3w3JXibIQ:-L-95oNiPrNkedWqDUxK:bnvkk79) ([raw file](https://github.com/cockroachdb/cockroach/blob/f01fd8006ecacd715799e94f174d3010cb0943f9/c-deps/libroach/env_switching.cc#L22)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

```
\        /
  \    /
    \/
base_env
   (Env)
```

and a similar cone at the top. I had to scratch my head for a couple dozen seconds.
</blockquote></details>

Done. I used straight lines because it otherwise breaks single-line comments.
Switched to a multi-line comment instead and some pretty little boxes around each entity.

---

*[c-deps/libroach/env_switching.cc, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-8zWxV-xQkh6BD1ljI:-L-9I8ZR4oWd7vs1ylW3:b-2n46nn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f01fd8006ecacd715799e94f174d3010cb0943f9/c-deps/libroach/env_switching.cc#L30)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Do we ever call this without an env? If not, probably safer to disallow it.
</blockquote></details>

`options.env` is NULL by default. Currently, it is only set for in-memory DBs.

---

*[pkg/base/store_spec.go, line 69 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-910L3fwREQs8Tf1KC:-L-97lxaTWNEYRg9dICr:b-2x8yd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/base/store_spec.go#L69)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

What's happening here?
</blockquote></details>

uh. oops, old comment. Replaced with a relevant one.

---

*[pkg/base/store_spec.go, line 170 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-91APOh7DHOUGeut9E:-L-9ItO7I39kITsLmG3c:b-l3u21k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/base/store_spec.go#L170)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

```go
const path = "path"
```

and then use that twice below.
</blockquote></details>

Done and in the `EncryptionSpec` as well.

---

*[pkg/base/store_spec.go, line 69 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-94TNURzRtP3h4i_i0:-L-9IViTr3qKTmyzO1NB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L69)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Needs explanation.
</blockquote></details>

Done.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 39 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-91UmvKHphnYcZ5QVj:-L-96aDxEw90amd1tpp8:b-z4yr63) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/baseccl/encryption_spec.go#L39)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Why not https://golang.org/pkg/net/url/#URL and avoid a hand-rolled format (simply strip the `<Scheme>://`)? That avoids lots of subtleties with the characters allowed in paths and looks familiar to everyone.
</blockquote></details>

I'm following the StoreSpec. I want to keep the same behavior, but I especially want to make sure the parsing is exactly the same as we need to match paths.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 129 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-927iRr5S9iJMagc3i:-L-96oRaWw_EQrylt119:b3fdhio) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/baseccl/encryption_spec.go#L129)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I'm confused why this returns the hardcoded flag name.
</blockquote></details>

Ported over from `StoreSpec`. While I agree that referring to the flag variable would be best, I'm not a fan of the added dependency. `base` and `cli` (or their `ccl` variants) are already heavily depended upon. That said, we talking about `cliflags`, not `cli` so it's definitely doable.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 161 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-94eZb74edHDAdP_P2:-L-9K_S9eXXsH22bXkhT:bz7qjdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/ccl/baseccl/encryption_spec.go#L161)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The comment doesn't mention that we're going to manipulate the store spec list. Should this be called `MergeEncryptionSpecsIntoStoreSpecs`?
</blockquote></details>

Done. Renamed to `PopulateStoreSpecsWithEncryption` to make it clear what we're mutating.

---

*[pkg/ccl/cliccl/start.go, line 40 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-932nl6bmtBybka5kV:-L-9L05aqjuK-cEChxaj:b-9wlhfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/cliccl/start.go#L40)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Just do

```go
wrapped := cli.StartCmd.PersistentPreRunE
cli.StartCmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
  if err := wrapped(cmd, args); err != nil {
    return err
  }
  return matchStoreEncryptionSpecs(cmd, args)
}
```

and there is no need for `ExtraStartPreRunHook`.

To make this independent of init order, you can make both setters to `PersistentPreRunE` use this pattern.
</blockquote></details>

Good idea. Done here and in `cli/cli.go` for the start command.

---

*[pkg/ccl/cliccl/start.go, line 46 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-949-V1gGLQQkdpW7A:-L-9QGIouNMB3hQhreb-:btfnij6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/cliccl/start.go#L46)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Instead of exposing all of `ServerCfg`, just expose a getter into `serverCfg.Stores()` from `cli` that is clearly marked as something you shouldn't use in other contexts.
</blockquote></details>

Done with a warning.

---

*[pkg/cli/context.go, line 61 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-93y5SVGzHL6PA4d_Y:-L-9Oa1A6E35TMuO1hkd:bhmmzt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/cli/context.go#L61)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Not exclusively, I believe.
</blockquote></details>

ugh. Right you are. dropped the comment anyway now that it's no longer public.

---

*[pkg/storage/engine/enginepb/registry.proto, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-90Laewu-DQp83U2iR:-L-9NB0bA3K_aAj7yuj2:b-7txwzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fa0a4df33dd20e17eade406d3f2582970147095/pkg/storage/engine/enginepb/registry.proto#L40)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Add a comment specifying if this is a relative or absolute path. I suppose it's relative (to what?) to allow moving of mount points.
There is also the concern of moving a HDD from a Linux to a Windows machine. Will things break because the path separator changes? It might be better to store a repeated string with individual path components.
</blockquote></details>

Most likely relative but the Env doesn't get relative paths so I'll have to generate those myself.
As for the path separators, they are hard-coded to `/` in rocksdb. eg: this is how you get the MANIFEST path:
https://github.com/facebook/rocksdb/blob/master/util/filename.cc#L142 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-:-L-9RHYdzibade1vYUYv:b-p095gt)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 18 of 28 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/base/store_spec.go, line 73 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-7IR0AC778-MlT7fQq-r4-73:-L-9p7Vzzz8KDJX1seDu:bkj1ug) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L73)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Ack, I'll just ignore this part for now.
</blockquote></details>

@petermattis, @bdarnell: any thoughts on the best way to pass go-CCL objects blindly through `rocksdb.go::open()` and into libroach-CCL? I'll poke at trying to pass a C++ object that's only known in libroachccl, but I'm not sure go will be too happy about that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Without looking at this code: pass an encoded protobuf. I’ll take a look
tomorrow.

On Wed, Nov 29, 2017 at 9:54 PM marc <notifications@github.com> wrote:

> Review status: 18 of 28 files reviewed at latest revision, 15 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *pkg/base/store_spec.go, line 73 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-7IR0AC778-MlT7fQq-r4-73:-L-9p7Vzzz8KDJX1seDu:bkj1ug>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L73>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> Ack, I'll just ignore this part for now.
>
> @petermattis <https://github.com/petermattis>, @bdarnell
> <https://github.com/bdarnell>: any thoughts on the best way to pass
> go-CCL objects blindly through rocksdb.go::open() and into libroach-CCL?
> I'll poke at trying to pass a C++ object that's only known in libroachccl,
> but I'm not sure go will be too happy about that.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/20225>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/20225#issuecomment-348067204>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF6f94xz-eH27aGflGZ83ROKoJGXGMxyks5s7hjIgaJpZM4Qnm7_>
> .
>

mberhault commented:
That's one option and it would work fine for opaque config settings (eg: the encryption flag).

My other problem is figuring out how to have go-CCL setup cpp-CCL hooks into libroach for CCL-only code. Between overridden env and all the communication (stats reporting, periodic loops, etc..) we'll have a lot of CCL code spread between go and cpp that needs to communicate.

---

Review status: 18 of 28 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-:-L-BqV5cjvA_ZAosCFns:b-gfgjcx)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 11 of 11 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/db.cc, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-8yvExVJ2GAyq7TEjE:-L-CT-EKTLt3FImrp5HM:boh7use) ([raw file](https://github.com/cockroachdb/cockroach/blob/f01fd8006ecacd715799e94f174d3010cb0943f9/c-deps/libroach/db.cc#L101)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Done. I still need to figure out exactly who will own what. Right now, they're only passed to the `DBImpl` to be deleted properly. They're only used by the parent `DBEngine` through the options.
</blockquote></details>

Ok. I assume that'll be done by the end of this PR? Otherwise, please insert a TODO.

---

*[c-deps/libroach/env_switching.cc, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-8zdsu8Kw3w3JXibIQ:-L-CT4vntFvMrq8h5a8x:b74ws31) ([raw file](https://github.com/cockroachdb/cockroach/blob/f01fd8006ecacd715799e94f174d3010cb0943f9/c-deps/libroach/env_switching.cc#L22)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Done. I used straight lines because it otherwise breaks single-line comments.
Switched to a multi-line comment instead and some pretty little boxes around each entity.
</blockquote></details>

:heartbeat: 

---

*[pkg/base/store_spec.go, line 73 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-7IR0AC778-MlT7fQq-r4-73:-L-CTfN1H-mAJBECZLEY:bz6qvdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L73)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

@petermattis, @bdarnell: any thoughts on the best way to pass go-CCL objects blindly through `rocksdb.go::open()` and into libroach-CCL? I'll poke at trying to pass a C++ object that's only known in libroachccl, but I'm not sure go will be too happy about that.
</blockquote></details>

Like Pete said, a protobuf comes to mind. Or you can use a URL if human readable is preferred. `ear://<path>?<key>=<val>&<key>=<val>`.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 39 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-91UmvKHphnYcZ5QVj:-L-CU1D43JSeQyK5t1LP:bndldqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/baseccl/encryption_spec.go#L39)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I'm following the StoreSpec. I want to keep the same behavior, but I especially want to make sure the parsing is exactly the same as we need to match paths.
</blockquote></details>

Ah. That makes sense, but then it seems to makes sense to have a method that is called by both which parses the `string` into a `map[string]string`.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 129 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-927iRr5S9iJMagc3i:-L-CUTupO-Q_1tHNozQ1:b-qe2u0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/baseccl/encryption_spec.go#L129)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Ported over from `StoreSpec`. While I agree that referring to the flag variable would be best, I'm not a fan of the added dependency. `base` and `cli` (or their `ccl` variants) are already heavily depended upon. That said, we talking about `cliflags`, not `cli` so it's definitely doable.
</blockquote></details>

Hmm, `StoreSpec` also has this same problem. I have a mild preference for cleaning up both to refer to the variable name, if nothing else to highlight the dependency.

---

*[pkg/cli/context.go, line 70 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-CUyIRtzi3Pb8fbzKY:-L-CUyIRtzi3Pb8fbzKZ:bo6ej4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34cb90a08f536a2134e00bc54dd4c5e41d53dc8/pkg/cli/context.go#L70)):*
> ```Go
> // GetServerCfgStores provides direct public access to the StoreSpecList inside
> // serverCfg. This is used by CCL code to populate some fields.
> // WARNING: consider very carefully whether should should be using this.
> ```

should -> you and while you're here anyway, insert a newline before the warning.

---

*[pkg/cli/flags.go, line 164 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-CV48uG5ZbDlOJI7gY:-L-CV48uG5ZbDlOJI7gZ:b-8a7ihi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34cb90a08f536a2134e00bc54dd4c5e41d53dc8/pkg/cli/flags.go#L164)):*
> ```Go
> 	// The following only runs for `start`.
> 	// Make sure we save and call any existing PreRun hook as well.
> 	wrapped := StartCmd.PersistentPreRunE
> ```

Optional nit: you could even factor this out,

```go
// AddPersistentPreRunE adds the given hook after any existing pre run hooks for the command.
// This allows hooks to be set without wiping out an existing hook.
func AddPersistentPreRunE(cmd *cobra.Command, hook func(*cobra.Command, []string) error) {
  wrapped := cmd.PersistentPreRunE
  cmd.PersistentPreRunE = func(c *cobra.Command, a []string) error {
    if wrapped != nil {
      if err := wrapped(c, a); err != nil { return err }
    }
    return hook(c, a)
  }
}
```

which makes it easy to grep for the "other" half of the hook setting and creates a distinguished place for explaining what's happening.

---

*[pkg/storage/engine/enginepb/registry.proto, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-90Laewu-DQp83U2iR:-L-CWMz3jKeAPJYRBisy:b-49d8vz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fa0a4df33dd20e17eade406d3f2582970147095/pkg/storage/engine/enginepb/registry.proto#L40)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Most likely relative but the Env doesn't get relative paths so I'll have to generate those myself.
As for the path separators, they are hard-coded to `/` in rocksdb. eg: this is how you get the MANIFEST path:
https://github.com/facebook/rocksdb/blob/master/util/filename.cc#L142 
</blockquote></details>

Interesting. And that just works on Windows? Worth making a note to test if nothing else.
If you're not ready to add a comment in this PR because it's still up in the air, please leave a TODO in its place.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/base/store_spec.go, line 73 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-7IR0AC778-MlT7fQq-r4-73:-L-Cq_UZ3Y-j-tAFh6FL:b3zm7vv) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L73)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Like Pete said, a protobuf comes to mind. Or you can use a URL if human readable is preferred. `ear://<path>?<key>=<val>&<key>=<val>`.
</blockquote></details>

For options, I would go with comma separated, semicolon separated or query-components (ampersand separated). I might not have enough context here to be giving good advice, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 4 of 28 files reviewed at latest revision, 9 unresolved discussions.

---

*[c-deps/libroach/db.cc, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-8yvExVJ2GAyq7TEjE:-L-Ccg7Eaeu8lBWfQK4V:b-mh20ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/f01fd8006ecacd715799e94f174d3010cb0943f9/c-deps/libroach/db.cc#L101)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Ok. I assume that'll be done by the end of this PR? Otherwise, please insert a TODO.
</blockquote></details>

Adding a TODO for now. Implementing the Switching Env is probably done separately.

---

*[pkg/base/store_spec.go, line 73 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-7IR0AC778-MlT7fQq-r4-73:-L-CcoBVUOK66VPrHCoy:b7n8e3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L73)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Like Pete said, a protobuf comes to mind. Or you can use a URL if human readable is preferred. `ear://<path>?<key>=<val>&<key>=<val>`.
</blockquote></details>

Yeah, my bigger concern is that a lot of the encryption objects on the C++ side will be CCL. This mean initializing them in go-ccl code, then passing them through somehow. I'm still investigating, so I'll take the `ExtraFields` out for now.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 39 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-91UmvKHphnYcZ5QVj:-L-Cd0sNna18uSR5bpMO:bov9np8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/baseccl/encryption_spec.go#L39)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Ah. That makes sense, but then it seems to makes sense to have a method that is called by both which parses the `string` into a `map[string]string`.
</blockquote></details>

True, hence why I'm sharing path parsing. Another kink is that the `StoreSpec` can have an optional path field when a single value is specified. Still, it makes sense to generalize the code. Putting in a TODO.

---

*[pkg/ccl/baseccl/encryption_spec.go, line 129 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-927iRr5S9iJMagc3i:-L-CitkfEmDOc1lod8SR:b1bar6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/ccl/baseccl/encryption_spec.go#L129)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Hmm, `StoreSpec` also has this same problem. I have a mild preference for cleaning up both to refer to the variable name, if nothing else to highlight the dependency.
</blockquote></details>

Changed the `StoreSpec` to use the `FlagInfo`. Added a TODO for the encryption spec. I'll refactor CCL flags to use `cliflags` right after this PR (it's a bunch of code movement, I'd rather keep it separate)

---

*[pkg/cli/context.go, line 70 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-CUyIRtzi3Pb8fbzKY:-L-CeU8jgd2PHEmnbrSj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34cb90a08f536a2134e00bc54dd4c5e41d53dc8/pkg/cli/context.go#L70)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

should -> you and while you're here anyway, insert a newline before the warning.
</blockquote></details>

Done.

---

*[pkg/cli/flags.go, line 164 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-CV48uG5ZbDlOJI7gY:-L-CozQ9xr_Vd_o3PSS8:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34cb90a08f536a2134e00bc54dd4c5e41d53dc8/pkg/cli/flags.go#L164)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Optional nit: you could even factor this out,

```go
// AddPersistentPreRunE adds the given hook after any existing pre run hooks for the command.
// This allows hooks to be set without wiping out an existing hook.
func AddPersistentPreRunE(cmd *cobra.Command, hook func(*cobra.Command, []string) error) {
  wrapped := cmd.PersistentPreRunE
  cmd.PersistentPreRunE = func(c *cobra.Command, a []string) error {
    if wrapped != nil {
      if err := wrapped(c, a); err != nil { return err }
    }
    return hook(c, a)
  }
}
```

which makes it easy to grep for the "other" half of the hook setting and creates a distinguished place for explaining what's happening.
</blockquote></details>

Good idea, done.

---

*[pkg/cli/start.go, line 654 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-94KT7ksT2cYdIYuzs:-L-Cp0vKQSmV5k9ZDy8w:b-p9iyxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cca4c154f0e34f43abb16a4ddc8912af86ed732/pkg/cli/start.go#L654)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Ignoring for now.
</blockquote></details>

I'm taking this out for now.

---

*[pkg/storage/engine/enginepb/registry.proto, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-90Laewu-DQp83U2iR:-L-CqE49lA8fC3jcWeqV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fa0a4df33dd20e17eade406d3f2582970147095/pkg/storage/engine/enginepb/registry.proto#L40)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Interesting. And that just works on Windows? Worth making a note to test if nothing else.
If you're not ready to add a comment in this PR because it's still up in the air, please leave a TODO in its place.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Ok, I'm removing the extra fields and will leave this to a subsequent PR. I'd like to merge this one as is, then refactor some of the cli flags and spec parsing.

---

Review status: 4 of 28 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-:-L-CrP1gp1CNf0DrUOHA:b-ucwrwy)*
<!-- Sent from Reviewable.io -->

benesch commented:
Drive by comment: Little known fact that Windows will handle forward
slashes in paths just fine, save for maybe some real obscure APIs. It’s
drive letters (ie, constructing absolute paths) that really cause trouble.

On Thu, Nov 30, 2017 at 11:56 AM Peter Mattis <notifications@github.com>
wrote:

> Review status: all files reviewed at latest revision, 8 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *pkg/base/store_spec.go, line 73 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-7IR0AC778-MlT7fQq-r4-73:-L-Cq_UZ3Y-j-tAFh6FL:b3zm7vv>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L73>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> Like Pete said, a protobuf comes to mind. Or you can use a URL if human
> readable is preferred. ear://<path>?<key>=<val>&<key>=<val>.
>
> For options, I would go with comma separated, semicolon separated or
> query-components (ampersand separated). I might not have enough context
> here to be giving good advice, though.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/20225>*
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/20225#issuecomment-348251162>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IGI1T0AvFQPSTfI3YJTEy9EHC7RMks5s7t5HgaJpZM4Qnm7_>
> .
>

tschottdorf commented:
LGTM

On Thu, Nov 30, 2017 at 12:01 PM Nikhil Benesch <notifications@github.com>
wrote:

> Drive by comment: Little known fact that Windows will handle forward
> slashes in paths just fine, save for maybe some real obscure APIs. It’s
> drive letters (ie, constructing absolute paths) that really cause trouble.
>
> On Thu, Nov 30, 2017 at 11:56 AM Peter Mattis <notifications@github.com>
> wrote:
>
> > Review status: all files reviewed at latest revision, 8 unresolved
> > discussions, all commit checks successful.
> > ------------------------------
> >
> > *pkg/base/store_spec.go, line 73 at r4
> > <
> https://reviewable.io:443/reviews/cockroachdb/cockroach/20225#-L-7IR0AC778-MlT7fQq-r4-73:-L-Cq_UZ3Y-j-tAFh6FL:b3zm7vv
> >
> > (raw file
> > <
> https://github.com/cockroachdb/cockroach/blob/42e6632b956ab718197529b9e828d85811fc4b4c/pkg/base/store_spec.go#L73
> >):*
> > *Previously, tschottdorf (Tobias Schottdorf) wrote…*
> >
> > Like Pete said, a protobuf comes to mind. Or you can use a URL if human
> > readable is preferred. ear://<path>?<key>=<val>&<key>=<val>.
> >
> > For options, I would go with comma separated, semicolon separated or
> > query-components (ampersand separated). I might not have enough context
> > here to be giving good advice, though.
> > ------------------------------
> >
> > *Comments from Reviewable
> > <https://reviewable.io:443/reviews/cockroachdb/cockroach/20225>*
> >
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/cockroachdb/cockroach/pull/20225#issuecomment-348251162
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AA15IGI1T0AvFQPSTfI3YJTEy9EHC7RMks5s7t5HgaJpZM4Qnm7_
> >
> > .
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/20225#issuecomment-348252556>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135ILDbBSUuV6FTRkKLr_snnW2hC08ks5s7t9OgaJpZM4Qnm7_>
> .
>
-- 

-- Tobias
